### PR TITLE
feat(external table): adapt external tables

### PR DESCRIPTION
**Pull request by zijiacj**
#### Which issue(s) this PR fixes:

Fixes #

#### Additional documentation e.g., usage docs, etc.:

```docs

```

### CODE REVIEW

File: libs/db-browser/src/main/java/com/oceanbase/tools/dbbrowser/schema/DBSchemaAccessor.java [hunk 1]
```diff
@@ -81,6 +81,20 @@ default List<String> showTables(String schemaName) {
      */
     List<DBObjectIdentity> listTables(String schemaName, String tableNameLike);
 
+    /**
+     * Show all external table names list in the specified schema
+     */
+    default List<String> showExternalTables(String schemaName) {
```
Comment (yhilmare) [inline libs/db-browser/src/main/java/com/oceanbase/tools/dbbrowser/schema/DBSchemaAccessor.java:87]:
```java

    /**
     * Get all table columns in the specified schema
     */
    Map<String, List<DBTableColumn>> listTableColumns(String schemaName, List<String> tableNames);

    /**
     * Get all table columns in the specified schema and table
     */
    List<DBTableColumn> listTableColumns(String schemeName, String tableName);

    /**
     * Get all table columns(hold only basic info) in the specified schema
     */
    Map<String, List<DBTableColumn>> listBasicTableColumns(String schemaName);

    /**
     * Get all table columns(hold only basic info) in the specified schema and table
     */
    List<DBTableColumn> listBasicTableColumns(String schemaName, String tableName);
```

should we add these methods for external table?

File: libs/db-browser/src/main/java/com/oceanbase/tools/dbbrowser/schema/DBSchemaAccessor.java [hunk 2]
```diff
@@ -81,6 +81,20 @@ default List<String> showTables(String schemaName) {
...
+        return showExternalTablesLike(schemaName, null);
+    }
+
+    List<String> showExternalTablesLike(String schemaName, String tableNameLike);
```
Comment (yhilmare) [inline libs/db-browser/src/main/java/com/oceanbase/tools/dbbrowser/schema/DBSchemaAccessor.java:91]: `    Map<String, List<DBTableColumn>> listBasicColumnsInfo(String schemaName);`
should we modify the implementation of this method for external table?

File: libs/db-browser/src/main/java/com/oceanbase/tools/dbbrowser/schema/DBSchemaAccessor.java [hunk 3]
```diff
@@ -81,6 +81,20 @@ default List<String> showTables(String schemaName) {
...
+
+    /**
+     * List all external table as BObjectIdentity in the specified schema
+     */
+    List<DBObjectIdentity> ListExternalTables(String schemaName, String tableNameLike);
```
Comment (yhilmare) [inline libs/db-browser/src/main/java/com/oceanbase/tools/dbbrowser/schema/DBSchemaAccessor.java:96]: how about index, constraints, partitions and etc.?

File: server/odc-server/src/main/java/com/oceanbase/odc/server/web/controller/v2/DBSchemaController.java [hunk 4]
```diff
@@ -44,14 +46,25 @@ public class DBSchemaController {
     @Autowired
     private TableService tableService;
 
+    /**
```
Comment (yhilmare) [inline server/odc-server/src/main/java/com/oceanbase/odc/server/web/controller/v2/DBSchemaController.java:49]: useless comments

File: server/odc-server/src/main/java/com/oceanbase/odc/server/web/controller/v2/DBTableController.java [hunk 5]
```diff
@@ -69,6 +69,14 @@ public ListResponse<String> listTables(@PathVariable String sessionId,
         }
     }
 
+    /**
```
Comment (yhilmare) [inline server/odc-server/src/main/java/com/oceanbase/odc/server/web/controller/v2/DBTableController.java:72]: useless comments

File: server/plugins/schema-plugin-ob-mysql/src/main/java/com/oceanbase/odc/plugin/schema/obmysql/OBMySQLExternalTableExtension.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+import com.oceanbase.tools.dbbrowser.schema.DBSchemaAccessor;
+
+/**
+ * @description:
+ * @author: zijia.cj
+ * @date: 2024/8/23 14:31
+ * @since: 4.3.3
+ */
+@Extension
+public class OBMySQLExternalTableExtension implements ExternalTableExtensionPoint {
```
Comment (yhilmare) [inline server/plugins/schema-plugin-ob-mysql/src/main/java/com/oceanbase/odc/plugin/schema/obmysql/OBMySQLExternalTableExtension.java:37]: what about other method?
how about ob-oracle, oracle and etc?

File: libs/db-browser/src/main/java/com/oceanbase/tools/dbbrowser/schema/DBSchemaAccessor.java (hunk 1 above)
Comment (zijiacj) [reply libs/db-browser/src/main/java/com/oceanbase/tools/dbbrowser/schema/DBSchemaAccessor.java:87]: The implementation of showtables is the default check basic_table, so I wrote it out separately, and other can be reused

File: libs/db-browser/src/main/java/com/oceanbase/tools/dbbrowser/schema/DBSchemaAccessor.java (hunk 2 above)
Comment (zijiacj) [reply libs/db-browser/src/main/java/com/oceanbase/tools/dbbrowser/schema/DBSchemaAccessor.java:91]: I think it's better to use map, but change the existing code

Comment (yhilmare) [reply libs/db-browser/src/main/java/com/oceanbase/tools/dbbrowser/schema/DBSchemaAccessor.java:91]: what I mean is that `listBasicColumnsInfo` is designed for table and view, should we do any refactor to make it adapt for external table?

Comment (zijiacj) [reply libs/db-browser/src/main/java/com/oceanbase/tools/dbbrowser/schema/DBSchemaAccessor.java:91]: ok

File: server/odc-server/src/main/java/com/oceanbase/odc/server/web/controller/v2/DBSchemaController.java (hunk 4 above)
Comment (zijiacj) [reply server/odc-server/src/main/java/com/oceanbase/odc/server/web/controller/v2/DBSchemaController.java:49]: ok

File: server/odc-server/src/main/java/com/oceanbase/odc/server/web/controller/v2/DBTableController.java (hunk 5 above)
Comment (zijiacj) [reply server/odc-server/src/main/java/com/oceanbase/odc/server/web/controller/v2/DBTableController.java:72]: ok

File: libs/db-browser/src/main/java/com/oceanbase/tools/dbbrowser/schema/DBSchemaAccessor.java (hunk 3 above)
Comment (zijiacj) [reply libs/db-browser/src/main/java/com/oceanbase/tools/dbbrowser/schema/DBSchemaAccessor.java:96]: Reuse the table implementation. For index, constraints, partitions and etc, there is no requirement that it must be a basic table

File: server/odc-service/src/main/java/com/oceanbase/odc/service/connection/table/TableService.java [hunk 7]
```diff
@@ -98,30 +103,71 @@ public class TableService {
...
-            if (authenticationFacade.currentUser().getOrganizationType() == OrganizationType.INDIVIDUAL) {
-                return latestTableNames.stream().map(tableName -> {
-                    Table table = new Table();
-                    table.setName(tableName);
-                    table.setAuthorizedPermissionTypes(new HashSet<>(DatabasePermissionType.all()));
-                    return table;
-                }).collect(Collectors.toList());
+            if (types == null || types.contains(DBObjectType.TABLE)) {
+                // add logic about table and make table.type==DBObjectType.TABLE
+                TableExtensionPoint point = SchemaPluginUtil.getTableExtension(dataSource.getDialectType());
```
Comment (yizhouxw) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/connection/table/TableService.java:115]: TableExtensionPoint point; `point` variable may rename to `tableExtention`

and what if get extension failed or get null ?
often we can juedge if feature is supported via extesion point exists or not.

File: server/odc-service/src/main/java/com/oceanbase/odc/service/connection/table/TableService.java [hunk 8]
```diff
@@ -98,30 +103,71 @@ public class TableService {
...
+                Set<String> latestTableNames = point.list(conn, database.getName())
+                        .stream().map(DBObjectIdentity::getName).collect(Collectors.toSet());
+                if (authenticationFacade.currentUser().getOrganizationType() == OrganizationType.INDIVIDUAL) {
```
Comment (yizhouxw) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/connection/table/TableService.java:118]: the diffrence for INDIVIDUAL or TEAM may extract to diffrent method.

current implementation too much dupliacted code.

File: server/odc-service/src/main/java/com/oceanbase/odc/service/connection/table/model/QueryTableParams.java [hunk 9]
```diff
@@ -32,5 +36,8 @@ public class QueryTableParams {
     private Long databaseId;
     @NotNull
     private Boolean includePermittedAction;
+    // table belonging to type in this collection needs to be fetched. if null, only fetch collection of
```
Comment (yizhouxw) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/connection/table/model/QueryTableParams.java:39]: use multile line comment for field, so as others

File: server/odc-service/src/main/java/com/oceanbase/odc/service/session/OdcStatementCallBack.java [hunk 10]
```diff
@@ -336,6 +345,45 @@ private List<JdbcGeneralResult> consumeStatement(Statement statement, SqlTuple s
         return executeResults;
     }
 
+    private void checkIsExternalTalbe(JdbcGeneralResult jdbcGeneralResult) {
+        if (dialectType == DialectType.OB_MYSQL || dialectType == DialectType.OB_ORACLE) {
+            String obVersion = ConnectionSessionUtil.getVersion(connectionSession);
+            if (VersionUtils.isGreaterThanOrEqualsTo(obVersion, support_External_Table_Version_Number_For_Ob)) {
+                JdbcColumnMetaData jdbcColumnMetaData = null;
+                try {
+                    jdbcColumnMetaData = jdbcGeneralResult.getQueryResult()
```
Comment (yizhouxw) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/session/OdcStatementCallBack.java:354]: how about join query with base table and external table, 
if judge based on first column, the detect logic may not strong

File: server/odc-service/src/main/java/com/oceanbase/odc/service/session/OdcStatementCallBack.java [hunk 11]
```diff
@@ -336,6 +345,45 @@ private List<JdbcGeneralResult> consumeStatement(Statement statement, SqlTuple s
...
+                            .getMetaData().getColumns()
+                            .get(0);
+                } catch (Exception e) {
+                    throw new RuntimeException(e);
```
Comment (yizhouxw) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/session/OdcStatementCallBack.java:358]: here may not throw exception, just regard as no external table, or the query will result in failed

File: server/odc-service/src/main/java/com/oceanbase/odc/service/session/OdcStatementCallBack.java [hunk 12]
```diff
@@ -336,6 +345,45 @@ private List<JdbcGeneralResult> consumeStatement(Statement statement, SqlTuple s
...
+                    externalTables = schema2ExternalTables.get(catalogName);
+                    if (externalTables.stream().filter(
+                            name -> name.equals(tableName))
+                            .collect(Collectors.toList()).size() > 0) {
+                        jdbcGeneralResult.setIsExternalTable(true);
+                    }
+                } else {
+                    DBSchemaAccessor schemaAccessor = DBSchemaAccessors.create(connectionSession);
+                    externalTables = schemaAccessor.showExternalTables(catalogName);
+                    if (externalTables.stream().filter(
```
Comment (yizhouxw) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/session/OdcStatementCallBack.java:376]: may simplify use ColletionUtils.contains

File: server/odc-service/src/main/java/com/oceanbase/odc/service/session/model/SqlExecuteResult.java [hunk 13]
```diff
@@ -107,6 +107,7 @@ public class SqlExecuteResult {
     private boolean withFullLinkTrace = false;
     private String traceEmptyReason;
     private boolean withQueryProfile;
+    private Boolean isExternalTable;
```
Comment (yizhouxw) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/session/model/SqlExecuteResult.java:110]: no `is` prefix for boolean, and here may `containsExternalTable`

File: server/odc-service/src/main/java/com/oceanbase/odc/service/connection/table/TableService.java [hunk 14]
```diff
@@ -98,44 +105,91 @@ public class TableService {
     @Transactional(rollbackFor = Exception.class)
     @SkipAuthorize("permission check inside")
     public List<Table> list(@NonNull @Valid QueryTableParams params) throws SQLException, InterruptedException {
+        List<DBObjectType> types = params.getTypes();
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/connection/table/TableService.java:108]: we should return both basic table and external table when `param.getTypes()` returns null or empty collection.

return basic table when `param.getTypes()` returns only basic Table
return external table when `param.getTypes()` returns only external table

File: server/odc-service/src/main/java/com/oceanbase/odc/service/connection/table/model/QueryTableParams.java [hunk 15]
```diff
@@ -32,5 +36,10 @@ public class QueryTableParams {
     private Long databaseId;
     @NotNull
     private Boolean includePermittedAction;
+    /**
+     * table belonging to type in this collection needs to be fetched. if null, only fetch collection of
+     * basic table
+     */
+    private List<DBObjectType> types;
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/connection/table/model/QueryTableParams.java:43]: we may give it a default value~

File: server/odc-service/src/main/java/com/oceanbase/odc/service/session/OdcStatementCallBack.java [hunk 16]
```diff
@@ -123,6 +128,8 @@ public class OdcStatementCallBack implements StatementCallback<List<JdbcGeneralR
     private final AsyncExecuteContext context;
     private final List<SqlExecutionListener> listeners = new ArrayList<>();
     private final ExecutorService executor = Executors.newSingleThreadExecutor();
+    private Map<String, List<String>> schema2ExternalTables;
+    public static final String support_External_Table_Version_Number_For_Ob = "4.3.2";
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/session/OdcStatementCallBack.java:132]: support_External_Table_Version_Number_For_Ob -> MIN_OB_VERSION_FOR_EXTERNAL_TABLE

File: server/odc-service/src/main/java/com/oceanbase/odc/service/session/OdcStatementCallBack.java [hunk 17]
```diff
@@ -312,8 +319,10 @@ private List<JdbcGeneralResult> consumeStatement(Statement statement, SqlTuple s
                     jdbcGeneralResult.setTraceId(execDetails.getTraceId());
                     jdbcGeneralResult.setWithFullLinkTrace(execDetails.isWithFullLinkTrace());
                     jdbcGeneralResult.setTraceEmptyReason(execDetails.getTraceEmptyReason());
+                    checkIsExternalTalbe(jdbcGeneralResult);
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/session/OdcStatementCallBack.java:322]: typo

File: server/odc-service/src/main/java/com/oceanbase/odc/service/session/OdcStatementCallBack.java [hunk 18]
```diff
@@ -312,8 +319,10 @@ private List<JdbcGeneralResult> consumeStatement(Statement statement, SqlTuple s
...
                 });
             }
+
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/session/OdcStatementCallBack.java:325]: useless empty line

File: server/odc-service/src/main/java/com/oceanbase/odc/service/session/OdcStatementCallBack.java [hunk 19]
```diff
@@ -336,6 +345,45 @@ private List<JdbcGeneralResult> consumeStatement(Statement statement, SqlTuple s
         return executeResults;
     }
 
+    private void checkIsExternalTalbe(JdbcGeneralResult jdbcGeneralResult) {
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/session/OdcStatementCallBack.java:348]: don't do this in the class, you may do this in `SqlExecuteResult#initEditableInfo`

File: server/plugins/schema-plugin-api/src/main/java/com/oceanbase/odc/plugin/schema/api/ExternalTableExtensionPoint.java [hunk 20]
```diff
@@ -0,0 +1,33 @@
...
+
+import com.oceanbase.tools.dbbrowser.model.DBObjectIdentity;
+
+/**
+ * @description:
+ * @author: zijia.cj
+ * @date: 2024/8/23 14:51
+ * @since: 4.3.3
+ */
+public interface ExternalTableExtensionPoint extends ExtensionPoint {
```
Comment (yhilmare) [inline server/plugins/schema-plugin-api/src/main/java/com/oceanbase/odc/plugin/schema/api/ExternalTableExtensionPoint.java:31]: there is no need to add a new extension point, we may add a new parameter on  `TableEntensionPoint#list(..., @NonNull List<DBObjectType> types)`

**Comment by ysjemmm** [comment]
good

File: libs/db-browser/src/main/java/com/oceanbase/tools/dbbrowser/schema/mysql/MySQLNoLessThan5700SchemaAccessor.java [hunk 21]
```diff
@@ -221,6 +221,27 @@ public List<DBObjectIdentity> listTables(String schemaName, String tableNameLike
...
+    }
+
+    @Override
+    public List<DBObjectIdentity> listExternalTables(String schemaName, String tableNameLike) {
+        throw new UnsupportedOperationException("Not supported yet");
+    }
+
+    @Override
+    public Boolean isExternalTable(String schemaName, String tableName) {
+        return null;
```
Comment (yhilmare) [inline libs/db-browser/src/main/java/com/oceanbase/tools/dbbrowser/schema/mysql/MySQLNoLessThan5700SchemaAccessor.java:242]: return false may be better, same problem as others

File: server/odc-service/src/main/java/com/oceanbase/odc/service/connection/logicaldatabase/LogicalTableService.java [hunk 22]
```diff
@@ -261,8 +261,8 @@ public DetailLogicalTableResp detail(@NotNull Long logicalDatabaseId, @NotNull L
         TableExtensionPoint tableExtensionPoint =
                 SchemaPluginUtil.getTableExtension(logicalDatabase.getConnectType().getDialectType());
         if (tableExtensionPoint == null) {
-            throw new UnsupportedOperationException(
-                    "Unsupported dialect " + logicalDatabase.getConnectType().getDialectType());
+            throw new UnsupportedOperationException("the dialect " + logicalDatabase.getConnectType().getDialectType()
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/connection/logicaldatabase/LogicalTableService.java:264]: this exception message may be reverted

File: server/odc-service/src/main/java/com/oceanbase/odc/service/db/DBTableService.java [hunk 23]
```diff
@@ -88,16 +89,24 @@ public List<String> showTablesLike(@NotNull ConnectionSession session, String sc
...
+                            .collect(Collectors.toList()).size() > 0);
+        }
+        if (type == DBObjectType.EXTERNAL_TABLE) {
+            PreConditions.validExists(ResourceType.OB_TABLE, "tableName", tableName,
+                    () -> schemaAccessor.showExternalTables(schemaName).stream().filter(name -> name.equals(tableName))
+                            .collect(Collectors.toList()).size() > 0);
+        }
         try {
-            return connectionSession.getSyncJdbcExecutor(
+            DBTable table = connectionSession.getSyncJdbcExecutor(
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/db/DBTableService.java:105]: we may revert this change

File: server/odc-service/src/main/java/com/oceanbase/odc/service/db/schema/syncer/column/DBTableAndViewAndExternalTableColumnSyncer.java [hunk 24]
```diff
@@ -48,7 +48,7 @@ Map<String, Set<String>> getLatestObjectToColumns(@NonNull ColumnExtensionPoint
 
     @Override
     public Collection<DBObjectType> getColumnRelatedObjectTypes() {
-        return Arrays.asList(DBObjectType.TABLE, DBObjectType.VIEW);
+        return Arrays.asList(DBObjectType.TABLE, DBObjectType.VIEW, DBObjectType.EXTERNAL_TABLE);
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/db/schema/syncer/column/DBTableAndViewAndExternalTableColumnSyncer.java:51]: can we get external table's column by this way?
```java
Map<String, List<DBTableColumn>> listBasicTableColumns(String schemaName);
```

File: server/odc-service/src/main/java/com/oceanbase/odc/service/db/schema/syncer/object/DBExternalTableSyncer.java [hunk 25]
```diff
@@ -0,0 +1,78 @@
...
+    public DBObjectType getObjectType() {
+        return DBObjectType.EXTERNAL_TABLE;
+    }
+
+    @Override
+    Set<String> getLatestObjectNames(@NonNull TableExtensionPoint extensionPoint,
+            @NonNull Connection connection, @NonNull Database database) {
+        return extensionPoint.list(connection, database.getName(), DBObjectType.EXTERNAL_TABLE).stream()
+                .map(DBObjectIdentity::getName)
+                .collect(Collectors.toSet());
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/db/schema/syncer/object/DBExternalTableSyncer.java:71]: useless breaking lines

File: server/odc-service/src/main/java/com/oceanbase/odc/service/partitionplan/PartitionPlanService.java [hunk 26]
```diff
@@ -118,7 +119,8 @@ public List<DataType> getPartitionKeyDataTypes(@NonNull String sessionId,
         DialectType dialectType = connectionSession.getDialectType();
         TableExtensionPoint tableExtensionPoint = SchemaPluginUtil.getTableExtension(dialectType);
         if (tableExtensionPoint == null) {
-            throw new UnsupportedOperationException("Unsupported dialect " + dialectType);
+            throw new UnsupportedOperationException("the dialect " + connectionSession.getConnectType().getDialectType()
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/partitionplan/PartitionPlanService.java:122]: we may revert this change, same as below

File: libs/db-browser/src/main/resources/schema/sql/obmysql/obmysql_4_0_x.yaml [hunk 27]
```diff
@@ -29,7 +29,7 @@ sqls:
         FROM
           information_schema.tables
         WHERE
-          TABLE_SCHEMA = ? AND TABLE_TYPE = 'BASE TABLE'
+          TABLE_SCHEMA = ?
```
Comment (yhilmare) [inline libs/db-browser/src/main/resources/schema/sql/obmysql/obmysql_4_0_x.yaml:32]: why should we do this change?

**Review by yhilmare — APPROVED** [review]
lgtm

**Review by LuckyPickleZZ — APPROVED** [review]

**Review by guowl3 — APPROVED** [review]
lgtm

**Review by MarkPotato777 — APPROVED** [review]

**Review by yizhouxw — APPROVED** [review]
